okhttp/src/main/java/okhttp3/internal/Util.kt

```diff
@@ -0,0 +1,642 @@
+      isDaemon = daemon
+    }
+  }
+
+  /**
+   * Returns an array containing only elements found in [first] and also in [second].
+   * The returned elements are in the same order as in [first].
+   */
+  fun intersect(
+    comparator: Comparator<in String>,
+    first: Array<String>,
+    second: Array<String>
+  ): Array<String> {
+    val result = ArrayList<String>()
+    for (a in first) {
+      for (b in second) {
+        if (comparator.compare(a, b) == 0) {
+          result.add(a)
+          break
+        }
+      }
+    }
+    return result.toTypedArray()
+  }
+
+  /**
+   * Returns true if there is an element in [first] that is also in [second]. This
+   * method terminates if any intersection is found. The sizes of both arguments are assumed to be
+   * so small, and the likelihood of an intersection so great, that it is not worth the CPU cost of
+   * sorting or the memory cost of hashing.
+   */
+  fun nonEmptyIntersection(
+    comparator: Comparator<String>,
+    first: Array<String>?,
+    second: Array<String>?
+  ): Boolean {
+    if (first == null || second == null || first.isEmpty() || second.isEmpty()) {
+      return false
+    }
+    for (a in first) {
+      for (b in second) {
+        if (comparator.compare(a, b) == 0) {
+          return true
+        }
+      }
+    }
+    return false
+  }
+
+  fun hostHeader(url: HttpUrl, includeDefaultPort: Boolean): String {
+    val host = if (url.host().contains(":"))
+      "[" + url.host() + "]"
+    else
+      url.host()
+    return if (includeDefaultPort || url.port() != HttpUrl.defaultPort(url.scheme()))
+      host + ":" + url.port()
+    else
+      host
+  }
+
+  /**
+   * Returns true if [e] is due to a firmware bug fixed after Android 4.2.2.
+   * https://code.google.com/p/android/issues/detail?id=54072
+   */
+  fun isAndroidGetsocknameError(e: AssertionError): Boolean {
+    return (e.cause != null &&
+        e.message?.contains("getsockname failed") == true)
+  }
+
+  fun indexOf(comparator: Comparator<String>, array: Array<String>, value: String): Int {
+    var i = 0
+    val size = array.size
+    while (i < size) {
+      if (comparator.compare(array[i], value) == 0) return i
+      i++
+    }
+    return -1
+  }
+
+  @Suppress("UNCHECKED_CAST") fun concat(array: Array<String>, value: String): Array<String> {
+    val result = array.copyOf(array.size + 1)
+    result[result.size - 1] = value
+    return result as Array<String>
+  }
+
+  /**
+   * Increments [pos] until [input] is not ASCII whitespace. Stops at [limit].
+   */
+  fun skipLeadingAsciiWhitespace(input: String, pos: Int, limit: Int): Int {
+    for (i in pos until limit) {
+      when (input[i]) {
+        '\t', '\n', '\u000C', '\r', ' ' -> Unit
+        else -> return i
+      }
+    }
+    return limit
+  }
+
+  /**
+   * Decrements [limit] until `input[limit - 1]` is not ASCII whitespace. Stops at [pos].
+   */
+  fun skipTrailingAsciiWhitespace(input: String, pos: Int, limit: Int): Int {
```

Comment:
same, `lastIndexOfNonWhitespace()`

Reply:
(though awkwardly, the regular String.lastIndexOf() can return `limit`, whereas this will never return that value)